sdks/java/extensions/euphoria/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/provider/CompositeProvider.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.euphoria.core.translate.provider;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import org.apache.beam.sdk.extensions.euphoria.core.annotation.stability.Experimental;
+import org.apache.beam.sdk.extensions.euphoria.core.client.operator.base.Operator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.OperatorTranslator;
+import org.apache.beam.sdk.extensions.euphoria.core.translate.TranslatorProvider;
+
+/**
+ * An implementation of {@link TranslatorProvider} which allows to stack other {@link
+ * TranslatorProvider TranslatorProviders} in order given on construction time.
+ */
+@Experimental
+public class CompositeProvider implements TranslatorProvider {
+
+  private final List<TranslatorProvider> orderedTranslatorsChain;
+
+  private CompositeProvider(List<TranslatorProvider> orderedTranslatorsChain) {
+    requireNonNull(orderedTranslatorsChain);
+    this.orderedTranslatorsChain = orderedTranslatorsChain;
```

Comment:
would it be worth wrapping the collection in ImmutableList?

Reply:
Yes, good idea.